오픈 마켓 [STEP 1] 아리,제리

leeari95 opened the pull request:
안녕하세요. 그린 @GREENOVER 

아리 @leeari95, 제리 @llghdud921 입니다!
이번 오픈마켓 프로젝트 리뷰어로 만나뵙게되어 영광입니다. 🙇🏻‍♀️

오픈마켓 STEP 1 너무 어려웠는데... 저희가 잘 구현했는지 모르겠네요. 🥲
저희가 아래 적은 내용 이외에 부족한 점이나 놓친 부분이 있다면 번거롭더라도 한번 더 짚어주시면 감사하겠습니다. 🙏🏻

## 고민했던 점

### 1. 단일 책임 원칙(Single responsibility principle)

- 한 타입이 하나의 역할만 할 수 있도록 설계에 많은 고민을 해보았습니다.

### 2. CodingKeys 활용

실제 네트워크에서 내려오는 변수명이 스네이크 케이스를 사용하는 변수는 `Codingkey`를 이용하여 parsing하는 key를 바꿔주었으며, 스네이크케이스를 사용하지 않는, 즉 타입의 변수명과 일치하면 rawValue를 명시할 필요가 없어 가독성을 위해 한 줄로 case를 합쳐주었습니다.

```swift
enum CodingKeys: String, CodingKey {
   case id, stock, name, thumbnail, currency, price, images, vendors
   case vendorID = "vendor_id"
   case bargainPrice = "bargain_price"
   case discountedPrice = "discounted_price"
   case createdAt = "created_at"
   case issuedAt = "issued_at"
}
```

### 3. NetworkManager와 Network

- Network하는 과정에서 역할마다 객체를 구분하여 구현하였습니다.
    - `Network` : dataTask()를 통해 SessionDataTask를 서버로 전송해 직접 네트워킹하는 객체
        
        ```swift
        func execute(request: URLRequest, completion: @escaping (Result<Data?, Error>) -> Void) {
                session.dataTask(with: request) { data, response, error in
                ...
        ```
        
    - `NetworkManager` : Network의 excute를 통해 data를 받아 decoding하는 fetch()를 가진 객체
        
        ```swift
        func fetch<T: Decodable>(request: URLRequest,
                                    decodingType: T.Type,
                                    completion: @escaping (Result<T, Error>) -> Void) {
                
             network.execute(request: request) { result in
        ```
        

### 4. Name Space

- 하드코딩을 개선하기 위해 enum 타입을 만들어 Address와 HTTPMethod의 값들을 분류해주었습니다.

### 5. Request, Response

- Request할 때, 그리고 Response하는 타입이 세부적으로 달라 ProductModification, ProductRegistration 등... 각 타입을 모두 구현하였습니다.

### 6. Overloading function

- 상품 삭제, 등록, 조회 등 여러가지 요청을 request 메소드 하나를 오버로딩을 활용하여 작성하였습니다.

### 7. Test Doubles

- 테스트 작성을 위해 의존성 주입을 활용하여 Mock, Stub 객체를 만들어 활용하였습니다.
- URLProtocol을 상속받은 클래스를 만들고 재정의를 해주었습니다.
    - 이 방법은 URLSession의 dataTask를 직접 Stub으로 만드는 방법도 있었지만, URLSessionDataTask를 채택한 타입에 init()을 정의하니 deprecated 경고가 떠서 이를 해결하기 위해 삭제 후 URLProtocol을 활용하는 방법으로 로직을 변경하였습니다.

## 궁금했던 점 / 조언이 필요한 것

- **비동기 메서드를 사용하는 동기 메서드**는 비동기 메서드 테스트로 진행해야할까요?
    - 저희가 긴가민가해서 NetworkManager를 테스트 할 때 비동기 테스트 시 필요한 메소드를 빼고 진행해보았는데, 실패를 해야하는데 성공을 하는 케이스가 나오더라구요. 그래서 일단 사용을 해주는게 맞는 것 같아서 그렇게 진행했는데, 이게 올바른건지 잘모르겠어요.
- URLProtocol을 활용하여 Test를 진행
    - [WWDC2018](https://developer.apple.com/videos/play/wwdc2018/417/) 를 참고하여 URLProtocol을 상속받은 MockProtocol을 구현하여 test를 진행했는데요.
    - 야곰닷넷 Test에 오동나무가 알려주신 네트워크 Test 방식과 구글에 나온 [다수의 방식](https://techblog.woowahan.com/2704/)은 StubURLSessionDataTask를 구현하는 방법이었습니다.
    - 하지만 [문제가 발생](https://github.com/leeari95/ios-open-market#1-3-Trouble-Shooting)하여 URLProtocol을 커스텀 하는 방식으로 바꾸었는데요. URLProtocol 개념이 낯설어서 저희가 제대로 구현을 한건지 또 현업에서 네트워킹 Test를  어떤 방식으로 구현하는지 궁금합니다!
     (문제에 대한 자세한 사항은 README에 **Trouble Shooting**부분에 `1번문항`을 참고 부탁드립니다!)
- Health Checker의 필요성을 모르겠습니다..
    - API 문서를 들여다보면 맨 위에 Health Checker가 있는데, 이걸 어디에 활용하면 좋을지 몰라서 따로 Request를 구현하진 않았는데요. 혹시... 보통 어디서 쓰이는지가 궁금해서 여쭤봅니다.
- 테스트 시 Request의 바디도 체크를 해야할까요?
    - 테스트할 때 Request의 바디가 잘 들어갔는지 체크를... 해야할까요? 하고싶었는데, 뭔가 테스트하기 애매한 것 같아서 과감히 제외하고 테스트를 해주긴 했습니다. 그렇지만 이게 맞는건지 잘 판단이 서질 않아서 그린의 조언이 필요합니다!

## 해결이 되지 않은 점

- 실제 API를 활용해서 서버에 요청을 보낼 때를 테스트해보진 못했습니다. 요청시 `identifier`가 필요한데요. 아직 배정받은 `identifier`가 없어서 임의로 하드코딩 상태로 남겨두었어요. 나중에 `identifier`를 배정받게 되면 수정할 예정인데, 이 부분은 파라미터로 전달받는게 좋을까요?

```swift
// DELET - 상품 삭제
func request(id: UInt, secret: String) -> URLRequest? {
    ...
		// 이부분이요!!!!!
    request.addValue("80c47530-58bb-11ec-bf7f-d188f1cd5f22", forHTTPHeaderField: "identifier")
    
    return request
}
```

leeari95 committed:
style: SwfitLint 라이브러리 추가

llghdud921 committed:
feat: Products 타입 추가 #1

leeari95 committed:
feat: Product 타입 추가 #1

llghdud921 committed:
chore: products 파일 Assests에 추가

leeari95 committed:
feat: 파싱해주는 타입 Parser을 추가 #1

llghdud921 committed:
test: ParsetTest 추가 #2

- Products 타입에 pages 타입 변경
- Parser내 encode 파라미터 추가
- Swiftlint exclude 추가

leeari95 committed:
test: product를 파싱하는 테스트 추가 #2

- 파싱 테스트를 위한 json 파일 추가
- Vendors, Image 타입 추가
- Product 타입을 리팩토링
- SwiftLint에 경고 제외할 문자열과 옵트인 룰 추가

llghdud921 committed:
feat: network class 타입 및 NetworkError 추가 #3

leeari95 committed:
feat: URL과 HTTPMethod를 담고있는 타입 추가 #3

llghdud921 committed:
feat: NetworkManager 추가 #3

leeari95 committed:
feat: 상품 리스트 조회시 필요한 request를 반환하는 메소드 추가 #3

- SwiftLint yml파일에 disabled_rules 추가

llghdud921 committed:
feat: 상품 상세 조회를 위한 request메서드 추가 #3

leeari95 committed:
feat: 상품의 Secret을 조회할 때 필요한 Request 반환하는 메소드 추가 #3

- NetworkError에 case 추가
- ParserError에 case, errorDescription 추가하고 네이밍 수정
- Parser에 Encoder, encode 메소드 추가
- 상품의 Secret을 조회할 때 encoding 해야하는 타입 Secret 생성

llghdud921 committed:
feat: 상품 삭제를 위한 request 메소드 추가 #3

leeari95 committed:
feat: 상품 수정을 위한 request 메소드 추가 #3

- API 기능 명세에 있는 Currency 타입 추가후 전반적으로 리팩토링
- 상품 수정에 필요한 타입 ProductModification 추가

llghdud921 committed:
feat: 상품 등록을 위한 request 추가 #3

- Data 타입 extension 추가
- 파일 그룹화
- httpbody 생성하는 createBody() 추가
- multipart form body 구성을 위한 multipartForm() 추가
- imageFile, productRegistration 타입 추가
- multipartform protocol 추가

leeari95 committed:
refactor: NetworkManager 내부 리팩토링 #4

- 줄바꿈 최소한으로 제거
- 가독성 향상을 위한 변수 대입
- private 메소드를 extension으로 분리
- 접근제어 추가

llghdud921 committed:
refactor: NetworkError 내부 리팩토링 및 ProductSecret 타입명 변경 #4

leeari95 committed:
refactor: MultipartForm 프로토콜을 Protocol 그룹으로 이동 #4

- NetworkError에서 statusCode의 연관값 타입을 수정
- Vendors의 들여쓰기를 수정

llghdud921 committed:
refactor: Products 내 CodingKeys 수정 #4

leeari95 committed:
test:  Network의 excute 메소드를 테스트하는 코드 추가 #5

- 의존성 주입을 위해서 Sessionable 프로토콜 추가
- Network의 session 타입을 Sessionable로 변경하고 이니셜라이저 추가
- URLSession 테스트를 위해 Sessionable 프로토콜 채택
- 테스트를 위한 MockSession, MockURLSessionDataTask 타입 추가
- Network의 excute 메소드 성공과 실패 케이스 작성

llghdud921 committed:
test: NetworkManager의 fetch를 테스트하는 코드 추가 #5

- 의존성 주입을 위해 Networkable, Parsable 프로토콜 추가
- NetworkManager의 Network, parser 타입을 Networkable, Parsable로 변경
- 테스트를 위한 MockParser 추가 및 Parserable 프로토콜 채택
- NetworkManager의 fetch() 성공과 실패 케이스 코드 작성

leeari95 committed:
test: URLProtocol을 재정의하여 테스트 코드를 전반적으로 수정 #5

- URLSessionDataTask는 init()을 정의하니 deprecated 경고가 떠서 삭제 후 로직 변경
- Sessionable 프로토콜을 활용하여 의존성 주입을 주었던 부분 제거, 관련된 Mock 객체 삭제
- 테스트 코드에서 DummyData, MockSession 대신에 새로 만든 Session을 활용하도록 수정
- Mock JSON 파일 활용을 위해 stubProduct 프로퍼티와 StubProduct 타입 추가

llghdud921 committed:
refactor: NetworkManager내 multipart image 경로 수정 #4

leeari95 committed:
test: 상품리스트 조회에 맞는 Request를 반환하는 테스트코드 추가 #5

- setUpWithError, tearDownWithError 메소드를 재정의하여 테스트 기본 세팅 리팩토링

llghdud921 committed:
test: 상품 상세 조회를 반환하는 request를 확인하는 테스트코드 추가 #5

leeari95 committed:
test: 상품 삭제 Secret 조회에 맞는 Request를 반환하는 테스트코드 추가 #5

llghdud921 committed:
test: 상품 삭제를 반환하는 request를 확인하는 테스트코드 추가 #5

leeari95 committed:
test: 상품 수정에 맞는 Request를 반환하는 테스트코드 추가 #5

llghdud921 committed:
test: 상품 등록을 반환하는 request를 확인하는 테스트코드 추가 #5

leeari95 committed:
refactor: multipartFormRequest 메소드 내부에 request의 헤더를 추가하는 부분 수정 #4

- JSON이 아니라 multipart/fome-data로 수정
- boundary 변수를 함수 내부에 추가하여 활용하도록 수정
- 테스트 파일 내부 그룹으로 분리

llghdud921 committed:
refactor: MockURLProtocol내 주석 제거 #4

leeari95 committed:
refactor: NetworkConstant를 Address와 HTTPMethod로 분리 #4

llghdud921 committed:
docs: Step1 README 작성

leeari95 committed:
docs: README.md 파일 수정

GREENOVER commented on OpenMarket/OpenMarket/Error/ParserError.swift:
```diff
@@ -0,0 +1,25 @@
+    case decoding
+    case encoding
+    
+    var errorDescription: String {
+        switch self {
+        case .assestNotfound:
+            return "에셋 데이터를 찾을 수 없습니다."
+        case .decoding:
+            return "디코딩을 하는 도중 에러가 발생했습니다."
+        case .encoding:
+            return "인코딩을 하는 도중 에러가 발생했습니다."
+        }
+    }
```
case에 string값으로 선언해주는것과 이렇게 별도로 errorDescription으로 computed property로 빼는건 어떤 차이가 있을까요?
🤔

leeari95 replied:
가독성의 차이라고 생각합니다.
rawValue로 문자열을 가져오느냐 errorDescription으로 문자열을 가져오느냐의 차이인 것 같습니다!

GREENOVER commented on OpenMarket/OpenMarket/Error/ParserError.swift:
```diff
@@ -0,0 +1,25 @@
+import Foundation
+
+enum ParserError: LocalizedError {
+    case assestNotfound
```
앗 오타인것 같습니다..!
`asset` 혹은 `assets`를 의도하셨던것 같은데 s가 많아서 헷갈릴만 했네요😅

leeari95 replied:
앗... 미처 못봤네요ㅠㅠ 발견해주셔서 감사합니다!!! 🙇🏻‍♀️

leeari95 replied:
앗.. 캐치해주셔서 감사합니다!

GREENOVER commented on OpenMarket/OpenMarket/Error/ParserError.swift:
```diff
@@ -0,0 +1,25 @@
+
+enum ParserError: LocalizedError {
+    case assestNotfound
+    case decoding
+    case encoding
```
에러 케이스 네이밍이 조금 더 명확하면 어떨까해요..!
결론적으론 디코딩/인코딩 실패라고 보이는데 `decodingFail`같은건 어떨까요? 🙋🏻
디코딩이라는 네이밍만으로는 parserError는 발생했지만 디코딩이 실패한건지 어떤건지 알 수가 없지 않을까 해서요..!
위의 `assetNotFound`처럼 명확한 네이밍은 어떨까 하는 의견입니다ㅎㅎ

leeari95 replied:
원래 뒤에 Error를 붙여주었는데 뭔가 중복되는 거 같아서 수정해주었었거든요.
근데 의견주신 Fail 좋은것 같아요! 한번 적극 반영해보도록 하겠습니다 😊
감사합니다! 🙏🏻

GREENOVER commented on OpenMarket/OpenMarket/Extension/Data+extension.swift:
```diff
@@ -0,0 +1,17 @@
+
+import Foundation
+
+extension Data {
+    mutating func append(_ string: String) {
+        guard let data = string.data(using: .utf8) else {
+            return
+        }
+        self.append(data)
+    }
+}
```
기본타입에다 익스텐션들을 아주 잘 사용하시네요!👍👍

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/HTTPMethod.swift:
```diff
@@ -0,0 +1,15 @@
+
+import Foundation
+
+enum HTTPMethod: String {
+    case get = "GET"
+    case post = "POST"
+    case patch = "PATCH"
+    case delete = "DELETE"
+}
```
크.. enum 타입으로 관리하는것 아주 좋네요!
이렇게 된다면 개발자의 타이핑 실수를 확실히 줄여줄 수 있겠어요👍
지금처럼 최대한 코드에서 string literal을 바로 사용하는걸 지양해보면 좋겠네요!

GREENOVER commented on OpenMarket/OpenMarket/Error/NetworkError.swift:
```diff
@@ -0,0 +1,25 @@
+import Foundation
+
+enum NetworkError: LocalizedError {
+    case responseCasting
```
조금 더 명확한 케이스 네이밍은 어떨까요?🤔

leeari95 replied:
음.. 이것도 위에서 의견주신거 처럼 Fail을 뒤에 붙여주려고 하는데 괜찮을까요??

GREENOVER replied:
네! 좋습니다👍

GREENOVER commented on OpenMarket/OpenMarket/Protocol/MultipartForm.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+protocol MultipartForm {
+    var dictionary: [String: Any?] { get }
+}
```
프로토콜로 가져가신 이유가 궁금합니다🙋🏻

leeari95 replied:
```swfit
// POST - 상품 등록
    func request<T: MultipartForm>(params: T, images: [ImageFile]) -> Result<URLRequest?, Error> {
        ...
    }
```
해당 메소드가 제네릭으로 선언되어있는데요.
들어오는 타입은 딕셔너리라는 프로퍼티를 사용할 수 있으면 좋겠다고 생각해서 프로토콜을 별도로 선언해주어 활용해보았습니다.

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/Network.swift:
```diff
@@ -0,0 +1,35 @@
+import Foundation
+
+final class Network: Networkable {
+    let session: URLSession
```
해당 프로퍼티를 외부에서 알 필요가 있을까요?🤔
아래 execute 메서드를 통해 객체간 소통하기에 session 프로퍼티는 숨겨져도 좋을것 같습니다🙋🏻

leeari95 replied:
접근제어 부여하면서 이부분을 빼먹은 것 같네요. 꼼꼼한 피드백 고맙습니다 🙇🏻‍♀️

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/Network.swift:
```diff
@@ -0,0 +1,35 @@
+final class Network: Networkable {
+    let session: URLSession
+    
+    init(session: URLSession = .shared) {
+        self.session = session
+    }
```
custom session이 될 여지도 있어 이니셜라이저를 구현해주신건가요~?🙋🏻
만약 의도하시는게 해당 타입 객체는 항상 shared session이라면 이니셜라이저가 필요할까 해서 여쭤봤습니다!
shared session과 custom session의 차이는 뭘까요?🤔

leeari95 replied:
네! 말씀 주신것 처럼 추후에 custom session이 들어올 수도 있어서 이 타입 외에도 다른 타입들도 이니셜라이저을 구현해주려고 했습니다.
차이점은 URLSession에 대해서 학습할 때 공부해보았었는데요. 간단하게 아래처럼 알고있습니다!

* shared session
    * 델리게이트 없이 간단한 비동기 요청을 할 때 사용
* custom session
    * 세션의 configuration을 설정해서 커스텀이 가능하다.

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/Network.swift:
```diff
@@ -0,0 +1,35 @@
+        self.session = session
+    }
+    
+    func execute(request: URLRequest, completion: @escaping (Result<Data?, Error>) -> Void) {
```
에러처리를 Result 타입으로 선택하시고 잘 구현해주셨네요👍
에러처리에 대표적으로 크게 3가지 정도가 있는걸로 기억하는데 Result 타입을 선택하신 이유가 궁금해집니다🤔

leeari95 replied:
Result를 사용하면 일단 에러처리를 하는 부분이 깔끔해지는 장점이 있어 선호하는 편입니다.
또, complection 핸들러 내부에서 에러처리를 해주기에도 Result가 적합하다고 생각합니다.

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/Address.swift:
```diff
@@ -0,0 +1,32 @@
+            return URL(string: Address.baseURL + "/\(id)" + "/\(secret)")
+        }
+    }
+}
```
구현 너무 깔끔하고 잘 나눠졌으며 빼먹은건 없어 보이네요👍
다만 어떤 Address인지 조금 더 명확하면 어떨까요?🙋🏻
네트워크 그룹안에 있어서 지금이야 알지만 사실 그룹 디렉토리 없이 코드에서 사용한다면 어떤 주소인지 불명확해보입니다😅
만약 나~아중에 오픈마켓에서 배송까지 하게되어 집 주소 타입을 만든다면 헷갈릴 여지가 있어보입니다!
(지금은 전혀 문제 없어요🤭)

leeari95 replied:
오.. 이부분은 네이밍을 명확하게 개선해보도록 하겠습니다! 😊

GREENOVER commented on OpenMarket/OpenMarket/Model/Response/Vendors.swift:
```diff
@@ -0,0 +1,21 @@
+    
+    enum CodingKeys: String, CodingKey {
+        case name, id
+        case createdAt = "created_at"
+        case issuedAt = "issued_at"
```
아마 아실수도 있는데 만약 스네이크 케이스에서 단순히 카멜 케이스로만 바꿔주는 코딩키 전략을 세우시게 된다면
`keyDecodingStrategy`와 `keyEncodingStrategy`와 같은 유용한 프로퍼티도 존재합니다~!
https://developer.apple.com/documentation/foundation/jsondecoder/2949119-keydecodingstrategy
https://developer.apple.com/documentation/foundation/jsonencoder/2949141-keyencodingstrategy/
여러분들의 프로젝트에서 사용되지 않아도되요!
어떤게 더 편리하고 유용한지 판단해보면 될것 같아요ㅎㅎ
제 판단에는 지금 모델 타입에서는 단순 네이밍 스타일 컨벤션 외에 아예 네이밍 자체가 변하는것도 많아서 소개해드린 방식은 적합하진 않아 보이긴 합니다!

leeari95 replied:
와우.. 이런 유용한 프로퍼티가 존재하다니.. 처음알았어요. 그린 덕분에 좋은거 알고가네요! 감사합니다! 🙏🏻

GREENOVER commented on OpenMarket/OpenMarket/Model/Response/Image.swift:
```diff
@@ -0,0 +1,22 @@
+    let id: Int
+    let url: String
+    let thumbnailURL: String
+    let succeed: Bool
```
해당 프로퍼티만 보고도 어떤 성공인지 알 수 있는 네이밍은 어떨까요?🙋🏻

leeari95 replied:
`uploadSucceed`라고 수정해보려고 하는데 괜찮을까요? 😁
API 문서를 참고했어요!
> 이미지 리사이징 및 업로드 성공 여부. 해당 값이 실패일 경우 상품정보에서 노출되지 않는다.

GREENOVER replied:
네 훨씬 좋네요!

GREENOVER commented on OpenMarket/OpenMarket/Protocol/Networkable.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+protocol Networkable {
+    func execute(request: URLRequest, completion: @escaping (Result<Data?, Error>) -> Void)
+}
```
프로토콜을 아주 잘 다루시는군요!🙌

GREENOVER commented on OpenMarket/OpenMarket/Protocol/Parserable.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+protocol Parserable {
```
어떤걸 Parserable하는건지 명확한 네이밍은 어떨까요?
예를들어 `JSONParserable`같은..!?
스타일 차이입니다만 저는 최대한 이후를 생각해봐서 함축적이 아니더라도 가장 잘 표현해보려고 하는 스타일을 가지고 있습니다🙋🏻

leeari95 replied:
그린의 의견 적극 반영해서 수정해보았습니다!

GREENOVER commented on OpenMarket/.swiftlint.yml:
```diff
@@ -0,0 +1,21 @@
+included:
+  - OpenMarket # 프로젝트 이름
+excluded: # 린트 과정에서 무시할 파일 경로. `included`보다 우선순위 높음
+  - OpenMarket/AppDelegate.swift
+  - OpenMarket/SceneDelegate.swift
+identifier_name:
+    min_length:
+    - 2  # only min_length
+    error: 4 # only error
+    excluded: # 제외할 문자열 목록 사용
+    - id
+    - URL
+    - url
+    - GlobalAPIKey
+disabled_rules: # 실행에서 제외할 룰 식별자들
+  - colon
+  - control_statement
+  - trailing_whitespace
+  - vertical_parameter_alignment
+opt_in_rules: # 일부 룰은 옵트 인 형태로 제공
+  - empty_count
```
오호 SwiftLint를 써보셨군요!👍👍
저는 현업에서 두가지를 혼합해서 사용하고 있는데 하나가 바로 SwiftFormat입니다!
Apple에서 만들었기에 한번 보시면 좋을것 같아요🙌
https://github.com/apple/swift-format

GREENOVER commented on OpenMarket/.swiftlint.yml:
```diff
@@ -0,0 +1,21 @@
+included:
+  - OpenMarket # 프로젝트 이름
+excluded: # 린트 과정에서 무시할 파일 경로. `included`보다 우선순위 높음
```
Pods도 보통 excluded 처리해주곤 합니다~!
나중에 외부 라이브러리들을 사용할때 한번 해보세요ㅎㅎ

leeari95 replied:
오! 맞아요! 그렇더라구요! 외부라이브러리를 쓰게되면 꼭 활용해보겠습니다!

GREENOVER commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -130,21 +342,91 @@
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "if test -d \"/opt/homebrew/bin/\"; then\n  PATH=\"/opt/homebrew/bin/:${PATH}\"\nfi\n\nexport PATH\n\nif which swiftlint >/dev/null; then\n  swiftlint\nelse\n  echo \"warning: SwiftLint not installed, download from https://github.com/realm/SwiftLint\"\nfi\n";
```
빌드 파서에 쉘 스크립트도 잘 적용해보셨군요!
추후에 코드가 많아지고 수동으로 컨벤션을 수정하기 힘들다면
SwiftLint Rule중에 Supports autocorrection라는 좋은 기능도 있습니다🚀
스크립트에 같이 붙여줘서 편하게 사용하실 수 있을거에요🙌

leeari95 replied:
오... 감사합니다!!! 한번 추가해봐야겠어요!

GREENOVER commented on OpenMarket/.swiftlint.yml:
```diff
@@ -0,0 +1,21 @@
+identifier_name:
+    min_length:
+    - 2  # only min_length
+    error: 4 # only error
```
이건 어떤 룰인가요~?🧐

leeari95 replied:
경고가 뜨는 최소 글자수를 세팅해준 룰입니다.
4글자 이하일 때 경고가 나와서 설정해주었습니다!

GREENOVER commented on OpenMarket/.swiftlint.yml:
```diff
@@ -0,0 +1,21 @@
+    - 2  # only min_length
+    error: 4 # only error
+    excluded: # 제외할 문자열 목록 사용
+    - id
```
ID는 괜찮나요??

leeari95 replied:
헛.. 그렇네요? ID도 추후 필요하다면 추가해봐도 괜찮을까요?😁

GREENOVER commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -319,6 +601,54 @@
+				CODE_SIGN_STYLE = Automatic;
+				CURRENT_PROJECT_VERSION = 1;
+				GENERATE_INFOPLIST_FILE = YES;
+				IPHONEOS_DEPLOYMENT_TARGET = 15.0;
```
테스트는 15.0에서만 가능한건가요?🙋🏻

leeari95 replied:
헛.. 테스트를 만들다가 타겟이 저렇게 설정이 된 것 같아요! 이 부분은 어디서 수정하는 건지 여쭤볼 수 있을까요? 
일단 타겟 들어가니까... 이렇게 나와서요!

![image](https://user-images.githubusercontent.com/75905803/148394174-d649a0a2-4779-4f11-9272-d4a78b1af450.png)

GREENOVER replied:
아! 아래 오픈마켓 테스트 타겟에서 보셔야됩니다~!

GREENOVER commented on OpenMarket/OpenMarket/Assets.xcassets/product.dataset/product.json:
```diff
@@ -0,0 +1,28 @@
+    "id": 16,
+    "vendor_id": 2,
+    "name": "팥빙수",
+    "thumbnail": "https://s3.ap-northeast-2.amazonaws.com/media.yagom-academy.kr/training-resources/2/thumb/8279f1c16c6111ecbf33d51698ab7ead.png",
```
워우... 궁금해서 열었다고 놀랬네요ㅋㅋ

GREENOVER commented on OpenMarket/OpenMarket/Model/Parser.swift:
```diff
@@ -0,0 +1,37 @@
+
+import UIKit
+
+struct Parser: Parserable {
```
이친구도 조금 더 명확한 `JSONParser`와 같은 네이밍은 어떨까요?🙋🏻

GREENOVER commented on OpenMarket/OpenMarket/Model/Response/Product.swift:
```diff
@@ -0,0 +1,33 @@
+    var thumbnail: String
+    var currency: Currency
+    var price: Int
+    var bargainPrice: Int
+    var discountedPrice: Int
```
이 프로퍼티들은 필수값인가요?🙋🏻

leeari95 replied:
API 문서를 확인해보니까 Response를 받아올 때 저 프로퍼티들을 모두 받아오는 것 같아요!

GREENOVER replied:
아아 모두 옵셔널이 아닌 required한 값들이었나요? 그게 맞다면 제가 잘못봤나봐요ㅎㅎ..
그렇다면 그렇게 API 명세대로 가면 될것 같아요!
다만 해당 데이터들은 옵셔널이 더 어울리는데 문서 명세가 좀 어색하네요ㅋㅋ

leeari95 replied:
흠.. API 문서 확인해보니 Response 쪽에는 required 표시가 따로 없네요 ㅎㅎ

GREENOVER replied:
명확히 나타나있지 않다면 안전히 지금처럼 가시죠!

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,180 @@
+        return "Boundary-\(UUID().uuidString)"
+    }
+    
+    init(network: Networkable = Network(), parser: Parserable = Parser()) {
+        self.network = network
+        self.parser = parser
+    }
```
`network`와 `parser` 프로퍼티가 다른 타입으로 대체될 수 있는걸까요?
프로토콜을 상속받아 Network와 Parser 타입을 구현해주신것 같은데 다른 타입으로 갈아껴질 수 있는 여지 때문에 
이니셜라이저가 필요한건가 궁금합니다🙋🏻
만약 기능추가나 다른 부가적인 수정사항이 있다면 별도 타입을 만들 여지가 있을까해서요.
현재 Network와 Parser로 충분하지 않나 하는 생각이 들었어요🤔
여러분의 의견을 여쭙고 싶습니다!

leeari95 replied:
이건 사실 그린하고 같은생각으로 이니셜라이저가 아니라 일 반프로퍼티로 기본값을 대입하고 있었는데요.
의존성 주입을 해주다보니 이니셜라이저가 필요하게 되서 개선해주게 되었습니다!

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,180 @@
+        self.parser = parser
+    }
+    
+    func fetch<T: Decodable>(request: URLRequest,
+                            decodingType: T.Type,
+                            completion: @escaping (Result<T, Error>) -> Void) {
```
이런 코드 컨벤션은 어떠실까요?
```swift
func fetch<T: Decodable>(
    request: URLRequest,
    decodingType: T.Type,
    completion: @escaping (Result<T, Error>) -> Void
) {
...
}
```
메서드명 및 파라미터 시그니쳐가 길어질 경우 효과적인 컨벤션이라 느껴집니다🙋🏻
실제로 이런 컨벤션이 많이 사용되고 있는것으로도 알고 있어요!

leeari95 replied:
오... Swift Github에서 보았던 컨벤션이네요. 한번 아래부분도 함께 개선해보도록 하겠습니다!

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,180 @@
+    func fetch<T: Decodable>(request: URLRequest,
+                            decodingType: T.Type,
+                            completion: @escaping (Result<T, Error>) -> Void) {
+        
```
해당 줄바꿈은 어떤 의미일까요?🤔
다른 메서드 내부구현 스타일과 조금 달라서요..!

leeari95 replied:
이부분 아래 그린이 조언주신대로 컨벤션을 통일하도록 결정했습니다! 피드백 감사합니다 😁

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,180 @@
+    }
+    
+    // GET - 상품 리스트 조회
+    func request(page: UInt, itemsPerPage: UInt) -> URLRequest? {
```
파라미터로 메서드를 구분지어주고 동일한 네이밍을 사용하면 추후 사용하는 부분에서 혼동의 여지는 없을까요?🤔

leeari95 replied:
저희도 테스트 코드 작성할 때 혼동이 와서 고민했던 부분이라서... 네이밍을 다시한번 고민해보도록 하겠습니다!

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,180 @@
+    
+    // PATCH - 상품 수정
+    func request<T: Encodable>(data: T, id: UInt) -> Result<URLRequest?, Error> {
+        guard let url = Address.product(id: id).url else {
+            return .failure(NetworkError.notFoundURL)
+        }
+        let encodingResult = parser.encode(object: data)
+        let encodeData: Data
+        
+        switch encodingResult {
+        case .success(let data):
+            encodeData = data
+        case .failure:
+            return .failure(ParserError.encoding)
+        }
+        var request = URLRequest(url: url)
+        
+        request.httpMethod = HTTPMethod.patch.rawValue
+        request.httpBody = encodeData
+        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
+        request.addValue("80c47530-58bb-11ec-bf7f-d188f1cd5f22", forHTTPHeaderField: "identifier")
+
+        return .success(request)
```
이 로직은 위 상품 삭제 Secret 상세 조회 내부 로직과 굉장히 유사해보이고 중복되는 부분이 대부분인데 혹시 조금 더 다듬어 볼 수 있을까요?🤔
만약 이와 반대되는 생각으로는 중복을 줄이는것보다 지금 하신 방법의 구현이 더 합리적이거나 가독측면에서도 유리하다라는 의견이 들 수 도 있을거 같긴해요.
의견이 있다면 꼭 말씀해주세요..!

leeari95 replied:
음.. 각 리퀘스트를 수정, 등록 등등에 따라 나눠주기 위해 메소드로 별도로 구현해준 것인데.. 
만약 중복되는 부분을 합치게 된다면 메소드 네이밍이 많이 고민될 거 같아요.. 혹시 이부분은 어떻게 생각하실까요...???
그리고 삭제와 수정이 받아오는 파라미터가 약간 차이가 있어서 한꺼번에 다 받아와야 할 수도 있을 것 같아요.

GREENOVER replied:
음.. 분명 중복을 많이 줄여볼 순 있겠지만 그러기 위해선 다른 부수 코드도 생기긴 합니다!
스타일의 차이긴 하지만 저는 최대한 중복은 줄일것 같아요!
네이밍은 고민하면 되는거라 큰 문제가 되지 않아보여요😃 로직만 보고 생각해보면 좋을듯 싶습니다!

leeari95 replied:
음.. 내일 같이 고민해보고 한번 리팩토링 진행해보도록 해보아할 것 같습니다.
늦은밤까지 답변 달아주셔서.. 정말 감사합니다 😊

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,180 @@
+}
+
+extension NetworkManager {
+    private func multipartFormRequest<T: MultipartForm>(url: URL, params: T, images: [ImageFile]) -> URLRequest {
```
`requestMultipartForm`의 네이밍은 어떤가요?
지금까지의 코드 스타일로는 동사+명사의 느낌으로 메서드명이 지어졌는데 그 스타일이 깨진것 같습니다!
혹시 다른 의도가 있는걸까요~?🙋🏻

GREENOVER replied:
추가로 혹시 SwiftLint 룰 적용 시 한라인에 워딩 수를 제한하는 룰도 많이 쓰입니다..!
메서드명과 파라미터 시그니쳐가 길어진다면 개발자들은 가로를 보는것보다 세로를 보는것이 더 편하기에..
서로 화면 해상도도 다 다르고ㅠㅠ..
그래서 저는 파라미터가 3개가 넘어가거나 메서드명이 길어질때도 많아 최대한 아래 스타일을 가져갑니다.
(그냥 이런게 있다~ 하고 봐주세요..!)
```swift
private func multipartFormRequest<T: MultipartForm>(
  url: URL,
  params: T,
  images: [ImageFile]
) -> URLRequest {
...
}
```
아 참고로 저는 2칸 인덴트를 사용합니다.
이것도 당연히 스타일차이입니다!
그냥 코드가 방대해지면 단락이 바뀌고 가로로 길어지는걸 싫어해서 그래요😅

leeari95 replied:
영어로 네이밍하기 정말 어렵네요.. 추천해주신 네이밍으로 개선해보도록 할게요. 의견주셔서 감사합니다!!! 😁🙏🏻

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,180 @@
+import UIKit
+
+struct NetworkManager {
+    let network: Networkable
+    let parser: Parserable
+    var baseBoundary: String {
+        return "Boundary-\(UUID().uuidString)"
+    }
```
이 프로퍼티들은 외부에서 알 필요가 있을까요?🙋🏻

leeari95 replied:
앗 이부분도.. 휴먼에러가 난 것 같습니다!!!! 😂

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,180 @@
+        return body
+    }
+    
+    private func convertedMultiPartForm(name: String, value: Any, boundary: String) -> Data {
```
메서드명을 직역해보자면 변환된 멀티파트폼인데 동사의 형태를 가져가는 메서드명의 기조와 조금 상이한것 같아요..!
지금은 프로퍼티명에 더 적합한 네이밍이 되었지 않나 합니다🤔
어떤 역할을 해주는지 좀 더 네이밍으로 담아볼 수 있을것 같아요🙋🏻
아.. 제가 영알못이라 만약 다른 뜻이 있다면 말씀해주세요!

leeari95 replied:
헉.. 다시한번 고민해봐야될 것 같네요. 개선해서 피드백 반영해보도록 하겠습니다 😊

GREENOVER commented on OpenMarket/OpenMarket/Model/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,180 @@
+    
+    private func convertedMultiPartForm(name: String, value: Any, boundary: String) -> Data {
+        var data = Data()
+        data.append("--\(boundary)\r\n")
+        data.append("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n")
+        data.append("\(value)\r\n")
```
최대한 String Literal을 사용하지 않으려는 방향중 일환으로
\r\n\r\n 이런 부분들도 만약 많아질때를 대비해 상수 처리도 할 수 있습니다.
물론 지금 여기서는 그런 부분이 거의 없어 무의미해보일순 있습니다🥲

leeari95 replied:
이 부분은 이전에 말씀드린 것 처럼 따로 enum타입으로 빼주도록 하겠습니다!!! 🤩🙌🏻

GREENOVER commented on OpenMarket/OpenMarketTests/Test Doubles/MockSession.swift:
```diff
@@ -0,0 +1,17 @@
+
+class MockSession {
+    static var session: URLSession {
+        let configuration = URLSessionConfiguration.ephemeral
```
해당 유형으로 만들어주신 이유가 궁금합니다🙋🏻

llghdud921 replied:
ephemeral은 쿠키나 세션 정보가 남아있지 않게 되는 설정으로 private하게 사용할 때 주로 사용합니다.
예시로 사파리, 크롬에서 시크릿모드를 만들때 사용한다고 합니다.

직접 말씀드린것과 같이 테스트에서는 Mock의 용도로 사용하여 쿠키나 세션 정보가 필요없으므로 해당 유형을 사용했습니다!!

GREENOVER commented on OpenMarket/OpenMarketTests/Test Doubles/MockURLProtocol.swift:
```diff
@@ -0,0 +1,42 @@
+class MockURLProtocol: URLProtocol {
+    static var mockURLs = [URL?: (error: Error?, data: Data?, response: HTTPURLResponse?)]()
+    
+    override class func canInit(with request: URLRequest) -> Bool {
+        return true
+    }
+    
+    override class func canonicalRequest(for request: URLRequest) -> URLRequest {
+        return request
+    }
```
이 메서드들은 어떤 역할을 해주며 실제로 오버라이드해서 사용해야되게 필요한가요?🙋🏻

llghdud921 replied:
- caninit()
   프로토콜 subclass가 지정된 request을 처리할 수 있는지 여부를 결정하는 것으로 프로토콜 subclass가 요청을 처리할 수 있으면 true이고 그렇지 않으면 false를 반환합니다!
    canInit()을 재정의하여 시스템이 우리에게 제공하는 모든 request에 관심이 있음을 나타낸다고 합니다.

- canonicalRequest()
   지정한 요청의 표준 버전을 반환한다고 하네요! 

위 두 메서드는 프로토콜을 서브클래스할 경우 필수적으로 오버라이드하는 부분으로 [공식문서](https://developer.apple.com/documentation/foundation/urlprotocol)에 나와있어서 구현을 해주었습니다.

GREENOVER commented on OpenMarket/OpenMarketTests/Test Doubles/MockURLProtocol.swift:
```diff
@@ -0,0 +1,42 @@
+        self.client?.urlProtocolDidFinishLoading(self)
+    }
+    
+    override func stopLoading() { }
```
오버라이드해주고 아무 구현이 없는것 같아요!
추후 추가되는건가요?
그게 아니라면 현재로써는 불필요한 코드로 보입니다🤔

llghdud921 replied:
아까 말씀드린대로 애플에서 구현없이 두고있어서 해당 코드 유지하도록 하겠습니다! ☺️

GREENOVER commented on README.md:
```diff
@@ -1,7 +1,213 @@
-## iOS 커리어 스타터 캠프
+# 오픈 마켓
```
와우 리드미 작성 진짜 세세하고 최고네요👍
끝까지 화이팅입니다🙌

GREENOVER commented on OpenMarket/OpenMarketTests/Test Doubles/StubProducts.swift:
```diff
@@ -0,0 +1,10 @@
+
+import Foundation
+
+struct StubProduct: Decodable {}
```
얘는 필요한 코드인가요?🤔

llghdud921 replied:
이 코드는 NetworkManager의 fetch()의 디코딩하는 부분을 테스트하기 위해 decodable은 타입이지만 올바르지 않은 가짜 객체를 넣어 실패 케이스을 의도적으로 구현하였습니다.
```swift
// test하기 위한 부분
let parsingResult = parser.decode(source: data, decodingType: decodingType)

// test code
func test_Fetch_Decode_failure() {
let response = HTTPURLResponse(url: url, statusCode: 200, httpVersion: nil, headerFields: nil)
...
let decodingtype = StubProduct.self
```

GREENOVER commented on OpenMarket/OpenMarketTests/ParserTests.swift:
```diff
@@ -0,0 +1,37 @@
+@testable import OpenMarket
+
+class ParserTests: XCTestCase {
+
+    let parser: Parser = Parser()
+    
+    func test_Products_parsing() {
+        let title = "products"
+        
+        let parsing = parser.decode(fileName: title, decodingType: Products.self)
+        let result = try? parsing.get()
```
여러분들은 어떤 기준으로 줄바꿈을 하고 계신가요~?🙋🏻

llghdud921 replied:
줄바꿈에 대한 실수가 있었습니다ㅎㅎ 수정하겠습니다!!

GREENOVER commented on OpenMarket/OpenMarketTests/NetworkTests.swift:
```diff
@@ -0,0 +1,70 @@
+    
+    func test_Success() {
+        // given
+        let url = URL(string: "testURL")!
+        let data = Data()
+        let response = HTTPURLResponse(url: url, statusCode: 200, httpVersion: nil, headerFields: nil)
+        MockURLProtocol.mockURLs = [url: (nil, data, response)]
+        let session = MockSession.session
+
+        let network = Network(session: session)
+        let request = URLRequest(url: url)
+        let expectation = XCTestExpectation(description: "네트워크 실행")
```
요런 아래와 중복되는 프로퍼티들은 전역적으로 빼서 setUpWithError와 tearDownWithError를 이용해 초기화를 해보면서
각 테스트 메서드에서 값만 갈아끼워보는것도 방법일것 같은데 어떠신가요?🙋🏻

llghdud921 replied:
네 공통적으로 사용하는 코드 부분은 전역적으로 분리하여 코드 수정했습니다 :)

GREENOVER commented on OpenMarket/OpenMarketTests/NetworkManagerTests.swift:
```diff
@@ -0,0 +1,203 @@
+        let request = URLRequest(url: url)
+        let decodingtype = Products.self
+        let expectation = XCTestExpectation(description: "네트워크 실행")
+        
+        sutNetworkManager?.fetch(request: request, decodingType: decodingtype) { result in
```
아주 사소하지만..ㅋㅋ 여기에는 when 주석이 없는것 같아요!

llghdud921 replied:
실수가 있었네요ㅎ 수정하겠습니다~~~~~~

GREENOVER reviewed (changes requested):
@leeari95 @llghdud921 
안녕하세요 아리 제리!!
이번 프로젝트에서 처음 뵙네요😁
2주간 잘 부탁드립니다🙇🏻
처음에 러닝커브도 있는 쉽지 않은 프로젝트인데 너무 상세하게 아주 잘 구현해주셔서 놀랐어요🤭
본의 아니게 재밌게(?) 리뷰하다보니 코멘트들이 다소.. 아니 매우 많아졌어요ㅠㅠ...
저도 캠퍼해봤던 입장이라 리뷰 코멘트가 많으면 좋긴하지만 반면 짜증도 났기에 누구보다 그 입장을 잘 알기에..
우선 죄송합니다😭
다만 많은 코멘트에서 거의 대다수는 수정 요청이 아닌 정말 궁금해서 의견을 묻는 요지들이 많습니다.
(코드와 아주 별개의 것도 있구요..🥲)
그럼 리뷰를 너그럽게 봐주시면 감사하겠습니다🙌
***
### 칭찬 드리고 싶은 부분
- SwiftLint 룰을 아주 잘 도입해보셨네요! 아주 유용할거에요 💯
- 파일 그룹핑이 아주 세세하게 잘 되어 있어 구조 파악에 편했습니다 💯
- 각 역할들을 최대한 잘게 쪼개 가독성이 좋았어요 💯
- 테스트 코드를 기가 막히게 디테일하게 짜셨군요 💯
- 리드미 작성 정말 나중에 보답 받으실거에요 💯
- 쉽지 않으셨을텐데 대체적으로 통신과 모델링 관련해서 잘 구현해주셨네요! 💯
***
### 고민하셨던 부분에 대한 저의 의견
1. 단일 책임 원칙(Single responsibility principle)
- 이 부분이 아주 잘 되어 있으신거 보고 사실 좀 놀랐습니다🤭
정말 객체지향을 잘 알고 설계하신것 같았어요. 이 부분은 고민하신만큼 성과가 나왔던 부분인것 같네요!
2. CodingKeys 활용
- 요 부분에 대해선 리뷰 코멘트로도 몇가지 남겨보았어요..!
스네이크 케이스의 변환만 카멜 케이스로 가져가고 싶다면 코딩키 전략을 사용해도 됩니다ㅎㅎ
또한 모델링에서 required 해야하는 부분인지도 남겨보았습니다!
(대다수의 자세한 코멘트는 이곳보다 코드 라인바이라인으로 남겨봤어요ㅎㅎ
혹시 이곳에 더 상세히 남기는게 더 편하시면 말씀해주세요!)
3. NetworkManager와 Network
- 궁금한 부분이 있습니다!
왜 나눠주게 되었는지 기준이 궁금해요 사실🙋🏻
테스트의 목적을 생각하신걸까요?
아 저는 의견이 달라서 여쭤보는건 아닙니다!
실제 통신은 결국 네트워크 매니저를 통해 이뤄지는데 `Networkable` 프로토콜과 그걸 따르는 `Network` 타입을 만들어 거쳐 사용함으로
얻는 이점으로 어떤걸 생각하셨는지 궁금해요~!
4. Name Space
- 정말 아주 최고에요👍
최대한 String Literal을 직접 코드에서 사용하는 지양하면 실수도 줄이고 개발자 에러를 줄일 수 있어서 저도 그렇게 사용합니다🙌
5. Request, Response
- 아주아주 잘 모델링되어 있는 부분 확인했습니다!
다만 궁금한게 파일 그룹명이 모호한것 같아요.
사실 그룹안에 들어있는 타입들이 상품을 위한것 같아서 이를 아우룰수 있는 폴더 네이밍은 어떨까 생각이 들었어요🤔
Request / Response의 그룹 네이밍만 봐서는 하위 파일들을 보지 않고선 유추하기가 힘들것 같다고 느껴졌습니다.
그 하위 파일들을 각 모델로 나누고 구현하신 부분은 좋은것 같습니다!
6. Overloading function
- 이 부분도 궁금한걸 리뷰 코멘트에 적긴했습니다🙋🏻
요약하자면 "개발자가 네이밍이 같아 파라미터로 구분지어줘야하는 구현에서 실수할 여지가 있지 않을까?" 였습니다🧐
7. Test Doubles
- 테스트 저도 약한 부분인데 두분이 너무 잘 구현해주셨더라구요!
Mock과 Stub의 사용도 적절해보이구요.
아마 오동나무가 테스트 더블 관련해서 강의를 만들었던(?) 맞는지 모르겠지만 아니라면 준비하고 있을거라 생각되고..!
그걸 보셨는가 싶지만 이와 별개로 저는 test double에 대해 개념을 이 레퍼런스로 잡았던 기억이 있어 공유드립니다🙋🏻
https://blog.pragmatists.com/test-doubles-fakes-mocks-and-stubs-1a7491dfa3da?gi=e7d32c619169
저는 현업에서 fake 구현을 가장 많이 사용해요.
stub 테스트는 결국 원하는 도출값을 내놓고 시작하기에 실제 통신을 타는지 어떤지 확인할때 모호합니다.
이에 fake 구현은 가장 실제 통신 및 로직을 다 처리해보지만 속은 fake인 그런 기능을 해주기에 어떻게 보면 가장 테스트 다운 테스트라고 볼 수 있어요.
한번 참고해보시면 도움이 되실거에요..!🙋🏻
***
### 궁금 혹은 조언 요청에 대한 저의 의견
1. 비동기 메서드를 사용하는 동기 메서드는 비동기 메서드 테스트로 진행해야할까요?
- 정말 죄송한데 제가 이 질문을 정확하게 이해를 못했어요ㅠ..
조금 더 구체적으로 예시 혹은 설명을 주시면 다시 답변을 드려보겠습니다!
백프로 이해가 되지 않은 상태에서 유추해서 답변 드리면 더 혼란만 드릴것 같아서요🥲
2. URLProtocol을 활용하여 Test를 진행
- 지금 같이 구현하는 방법은 저도 사실 처음봤어요!
왜냐면 저의 경우는 통신 관련해서도 대개 라이브러리를 쓰기에 URLSession 애플 라이브러리로 통신을 구현하는것은 캠프 이후로 많이 없습니다.
(보통 Moya, Alamofire를 쓰기에)
그렇기에 저의 현업에서도 두 라이브러리로는 테스트도 조금 더 수월하게 진행할 수 있습니다.
다만 외부 라이브러리라도 내부 구현은 애플 라이브러리 기조를 벗어날 수 없기에 상동하다고 생각합니다.
그래서 두분이 트러블슈팅에 잘 나타내주신것을 토대로 저도 학습을 해봤는데 지금 구현에 문제는 없다고 판단됩니다!
다만 해당 프로토콜을 채택하면서 사용되지 않는 메서드들도 오버라이드 한 부분이 있다고 보여서 리뷰 남겨봤습니다..!
3. Health Checker의 필요성을 모르겠습니다.
- 사실 구현하기 나름이겠지만 현재 두분이 구현하신 코드에선 필요하지 않습니다.
Health Checker는 보다 네트워크 상태 모니터링에 필요한 부분입니다.
지속적으로 서버에 신호를 보내 응답이 오는지에 대해 체크하는 부분을 구현하고 이를 통해 잘못된 응답이 온다면
잘못된 리퀘스트가 아니라면 사실 서버가 죽은거겠죠?
그럴경우 바로 즉각적인 대응을 해줄 필요가 있어 그럴 경우 보통 사용되는걸로 알고 있습니다.
다만 클라에서 저도 그렇고 이 헬스 체커를 사용해본 기억은 없네요!
(물론 이를 가지고 필요하다면 구현을 해줄 순 있겠지만요..!)
그렇기에 지금처럼 필요하신 부분만 API문서에서 가져가셔도 무방합니다🙌
4. 테스트 시 Request의 바디도 체크를 해야할까요?
- 음 제 의견은 굳이 안해도 될것 같습니다!
테스트 코드는 정말 스타일의 차이이긴 하지만 굳이 저거까지 체크해야될까? 라는 두분과 동일한 생각이 있습니다.
어차피 request body를 포함해 정상적으로 request 되었다면 그에 대한 응답을 확인하면 되는 부분일것 같아요.
***
### 해결되지 않은 부분에 대한 저의 의견
- 두가지 스탠스를 취할 수 있을것 같습니다🙋🏻
저 Identifier가 변치 않는것이라면 전역적으로 상수처리할수 있을것 같고
그게 아니라면 말씀하신것처럼 파라미터로 받을것 같습니다!
그런데 제가 알기론 보통 고유값일테니 상수로 받는게 어떨까 합니다!
***
두분 너무 고생많으셨어요..!
다시 한번 리뷰 코멘트가 너무 심각하게 많은것 같아 죄송스럽네요..😅
의견이 다른 부분은 바로바로 말씀해주세요ㅎㅎ
아 두분의 구현에서 다음 스텝으로 넘어가는데 지장은 없어 보입니다.
그렇지만 코멘트가 있다보니 한번 확인해보시고 수정할 부분은 수정 후 리뷰 재요청을
그게 아니면 머지 요청을 해주세요~!🚀

leeari95 committed:
refactor: 테스트 코드에 중복되는 부분 리팩토링 #6

llghdud921 committed:
refactor: test code내 주석 및 줄바꿈 수정 #6

leeari95 committed:
refactor: Image의 프로퍼티 네이밍을 명확하게 수정 #6

llghdud921 committed:
refator: Content-type,MultipartForm String값 상수로 정의 #6

- MultiPartForm 프로토콜을MultiPartFormProtocol로 변경
- MultipartForm, contentType enum 구현 및 관련 코드 수정
- 오버로딩된 request 함수명 명확하게 개선
- 들여쓰기 컨벤션 개선

leeari95 committed:
refactor: 에러의 케이스중 네이밍을 명확하게 개선 #6

llghdud921 committed:
refactor: Parser, Parsable에 JSON을 추가한 네이밍으로 개선 #6

leeari95 committed:
refactor: Network 프로퍼티에 접근제어 부여하여 개선 #6

llghdud921 committed:
refactor: Address를 APIAdress로 네이밍 개선 #6

leeari95 commented:
그린!! 와우... 이렇게 코멘트로 뚜둘겨 맞아본 것은 처음이라... 행복합니다. 😁 🍏
꼼꼼하게 봐주시고 피드백해주신 덕에 수정사항이 많네요!!!
정말 감사합니다! 수정사항 간단히 정리하자면 아래와 같습니다.
문제 없다면 머지 부탁드리겠습니다. 감사합니다! 
새해복 많이 받으세요~ 🙇🏻‍♀️
(테스트의 타겟, 린트 룰 추가 등은 다음 스텝에서 개선해서 수정해보도록 하겠습니다. 감사합니다!!!)

## 수정사항
* 테스트 코드에 중복되는 부분을 개선
  * 빠진 주석 및 줄바꿈을 수정
* Image의 프로퍼티 네이밍을 명확하게 수정
* 하드코딩 되어있는 문자열을 따로 enum 타입으로 빼주어 개선
* 에러의 네이밍을 명확하게 개선
* Parser, Parsable의 네이밍을 JSON을 덧붙혀 명확하게 개선
* 접근제어가 붙어있지 않은 프로퍼티에 접근제어를 추가
* Address의 네이밍을 명확하게 개선 (APIAddress)

GREENOVER commented on OpenMarket/OpenMarketTests/NetworkManagerTests.swift:
```diff
@@ -9,34 +9,39 @@ import XCTest
     var stubProducts = NSDataAsset(name: "products")!.data
     var stubProduct = NSDataAsset(name: "product")!.data
+    var sutNetworkManager: NetworkManager?
+    var sutURL: URL!
+    var sutData: Data!
+    var sutRequest: URLRequest!
```
force Unwrapping 타입말고 조금 더 안전하게 사용해볼 수 있을까요?
테스트 코드여서 실제 앱 환경에서 문제는 없겠지만 forece Unwrapping은 항상 앱 크래쉬의 주요 원인이 됩니다🥲

leeari95 replied:
테스트 코드여서 상관없다고 생각해서 느낌표를 붙여주었는데, 그린 피드백대로 안전하게 사용해보도록 하겠습니다! 😊

GREENOVER reviewed (approved):
변경 사항 확인했습니다!
한가지 부분에 코멘트를 추가해봤어요🥲
그 외 전부 이상없고 아주 잘 구현해주셨습니다!
스텝 1은 이정도면 충분할것 같아 어프로브 후 머지하겠습니다~
변경사항이 생기는 부분은 마이너하기에 스텝2를 하시면서 같이 보완해보면 될것 같습니다!
고생하셨습니다🙌

++ 아실 수 있지만 깃헙에서 리뷰 재요청을 할때 우측 상단에 Reviewer 탭에서 새로고침 버튼을 눌러주시면
저에게 깃헙으로 재요청 알림이 갑니다!
활용해보시면 편할거에요ㅎㅎ
![스크린샷 2022-01-07 오전 9 43 42](https://user-images.githubusercontent.com/72292617/148472884-07b04362-8589-42a6-b677-c62c489cb7b3.png)

p.s. 아 커밋 메시지 규칙에서 궁금한게 하나 있는데요..!
두분은 단순 네이밍만 변경되는 커밋에는 어떤 prefix를 달아주고 계신가요?
현재는 refactor를 달아주는것으로 보이는데 이유가 궁금합니다🙋🏻
